allen_oephys_to_nwb/allen_ecephys_interface.py

```diff
@@ -0,0 +1,244 @@
+        }
+        return input_schema
+
+    def __init__(self, **input_args):
+        super().__init__(**input_args)
+
+    def get_metadata_schema(self):
+        metadata_schema = get_base_schema()
+        metadata_schema['properties']['Ecephys'] = get_base_schema(tag='Ecephys')
+        metadata_schema['properties']['Ecephys']['properties']['Device_1'] = get_schema_from_hdmf_class(pynwb.device.Device)
+        metadata_schema['properties']['Ecephys']['properties']['Device_2'] = get_schema_from_hdmf_class(pynwb.device.Device)
+        metadata_schema['properties']['Ecephys']['properties']['ElectrodeGroup'] = get_schema_from_hdmf_class(pynwb.ecephys.ElectrodeGroup)
+        metadata_schema['properties']['Ecephys']['properties']['ElectricalSeries_raw'] = get_schema_from_hdmf_class(pynwb.ecephys.ElectricalSeries)
+        return metadata_schema
+
+    def get_metadata(self):
+        """Auto-fill as much of the metadata as possible."""
+        metadata = dict()
+
+        # Get metadata info from files
+        subjects_info_path = self.input_args['source_data']['path_subjects_info']
+        if Path(subjects_info_path).is_file():
+            with open(subjects_info_path, 'r') as inp:
+                subjects_info = json.load(inp)
+
+        subject_info = {
+            'subject_id': '',
+            'line': '',
+            'age': '',
+            'anesthesia': ''
+        }
+
+        session_identifier = str(uuid.uuid4())
+        if 'path_processed' in self.input_args['source_data']:
+            with h5py.File(self.input_args['source_data']['path_processed'], 'r') as f:
+                session_identifier = str(int(f['tid'][0]))
+                if np.isnan(f['aid'][0]):
+                    print(f"File {self.input_args['source_data']['path_processed']} does not have 'aid' key. Skipping it...")
+                else:
+                    subject_id = str(int(f['aid'][0]))
+                    subject_info = subjects_info[subject_id]
+                    subject_info['subject_id'] = subject_id
+
+        # File metadata
+        metadata['NWBFile'] = {
+            'session_description': 'session description',
+            'identifier': session_identifier,
+            'session_start_time': datetime.strptime('1900-01-01 00:00:00', '%Y-%m-%d %H:%M:%S'),
+            'pharmacology': subject_info['anesthesia'],
+        }
+
+        # Subject metadata
+        metadata['Subject'] = {
+            'subject_id': subject_info['subject_id'],
+            'genotype': subject_info['line'],
+            'age': subject_info['age']
+        }
+
+        # Ecephys metadata
+        metadata['Ecephys'] = dict()
+        metadata['Ecephys']['Device_1'] = {
+            'name': 'Device_ecephys'
+        }
+
+        metadata['Ecephys']['ElectrodeGroup'] = {
+            'name': 'ElectrodeGroup',
+            'description': 'no description',
+            'location': '',
+            'device': 'Device_ecephys'
+        }
```

Comment:
Can you use a single dict definition where possible?

Reply:
I don't understand, like what?

Reply:
```suggestion
        metadata = dict(
            Subject=dict(
                subject_id=subject_info['subject_id'],
                genotype=subject_info['line'],
                age=subject_info['age']
            ),
            Ecephys=dict(
                Device=dict(name='Device_ecephys'),
                ElectrodeGroup=dict(
                    name='ElectrodeGroup',
                    description='no description',
                    location='unknown',
                    device='Device_ecephys'
                )
            )
        )
```